File: src/main/java/com/tronic/bot/storage/Serializer.java
package com.tronic.bot.storage;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import com.toddway.shelf.Shelf;
import kotlin.jvm.JvmClassMappingKt;
import kotlin.reflect.KClass;
import org.jetbrains.annotations.NotNull;

import java.lang.reflect.Type;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;

public class Serializer implements Shelf.Serializer {
    private final Gson gson;

    public Serializer() {
        this.gson = new Gson();
    }

    @NotNull
    @Override
    public <T> String fromType(@NotNull T t) {
        return this.gson.toJson(t);
    }

    @NotNull
    @Override
    public <T> T toType(@NotNull String s, @NotNull KClass<T> kClass) {
        return this.gson.fromJson(s, JvmClassMappingKt.getJavaClass(kClass));
    }

    @NotNull
    @Override
    public <T> List<T> toTypeList(@NotNull String s, @NotNull KClass<T> kClass) {
        Class<T> cl = JvmClassMappingKt.getJavaClass(kClass);
        Type typeOfT = TypeToken.getParameterized(List.class, cl).getType();
        return this.gson.fromJson(s,typeOfT);
    }
}

File: src/main/java/com/tronic/bot/storage/Serializer.kt
package com.tronic.bot.storage;

import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken
import com.toddway.shelf.Shelf;
import kotlin.reflect.KClass;

class Serializer(private val gson:Gson = Gson()): Shelf.Serializer {

    override fun <T : Any> fromType(value: T): String {
        val dbg = this.gson.toJson(value)
        return dbg;
    }

    override fun <T : Any> toType(string: String, klass: KClass<T>): T {
        val dbg =  this.gson.fromJson<T>(string,klass::class.java)
        return dbg
    }

    override fun <T : Any> toTypeList(string: String, klass: KClass<T>): List<T> {
        val typeOfT = TypeToken.getParameterized(List::class.java, klass::class.java).type;
        return this.gson.fromJson(string,typeOfT);
    }

}
